### PR TITLE
Preliminary sdk config

### PR DESCRIPTION
**Pull request by mwear**
This PR is an implementation of SDK config for currently implemented features. It's based off of #165, but with speculative, future feature configuration removed. With this implementation you can configure logging, tracer factory, span processors and install instrumentation.

Here are some examples:

### Minimal setup

If a block is not specified it will setup defaults
* STDOUT logging
* SDK::Trace::TracerFactory
* SimpleSpanProcessor with ConsoleSpanExporter
* No instrumentation

```ruby
OpenTelemetry::SDK.configure
```

### Customizing the setup

The following can be customized:
* Logger
* Tracer Factory
* Span Processors
* Instrumentation: installed individually, or all, but not both

Customize the setup and specify instrumentation individually

```ruby
OpenTelemetry::SDK.configure do |c|
  c.tracer_factory = ...
  c.logger = Logger.new('/dev/null')
  c.add_span_processor(...)
  c.use 'OpenTelemetry::Adapters::Sinatra', {enabled: true}
  c.use 'OpenTelemetry::Adapters::Faraday', {ignore_filters: [/domain\.com/]}
  c.use 'SomeCustom::Adapter::Lib'
end
```

Customize the setup and install all instrumentation

```ruby
OpenTelemetry::SDK.configure do |c|
  c.tracer_factory = ...
  c.logger = Logger.new('/dev/null')
  c.add_span_processor(...)
  c.use_all 'OpenTelemetry::Adapters::Something' => {config_opt: true}
end
```

### CODE REVIEW

File: sdk/lib/opentelemetry/sdk.rb [hunk 1]
```diff
@@ -15,9 +15,52 @@
...
+    #   Add a span processor
+    #
+    #     OpenTelemetry::SDK.configure do |c|
+    #       c.add_span_processor SpanProcessor.new(SomeExporter.new)
+    #     end
+    #
+    #   Configure everything
+    #
+    #     OpenTelemetry::SDK.configure do |c|
+    #       c.tracer_factor = SomeTracerFactory.new
```
Comment (fbogsany) [inline sdk/lib/opentelemetry/sdk.rb]: s/tracer_factor/tracer_factory/

File: sdk/lib/opentelemetry/sdk.rb [hunk 2]
```diff
@@ -15,9 +15,52 @@
...
+    #     end
+    def configure
+      configurator = Configurator.new
+      yield configurator if block_given?
+      configurator.configure
+    end
   end
 end
 
+require 'opentelemetry/sdk/configurator'
```
Comment (fbogsany) [inline sdk/lib/opentelemetry/sdk.rb]: Naive question: should this `require` come before the `Configurator.new` or does it not matter (does the constant resolve happen at the point of call or at the point of compilation?).

File: sdk/lib/opentelemetry/sdk/configurator.rb [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+      def initialize
+        @adapter_names = []
+        @adapter_config_map = {}
+        @span_processors = []
+        @use_mode = USE_MODE_UNSPECIFIED
+      end
+
+      def tracer_factory
+        @tracer_factory ||= Trace::TracerFactory.new
```
Comment (fbogsany) [inline sdk/lib/opentelemetry/sdk/configurator.rb]: Is it necessary to support replacing the tracer factory? I can see you'd want to do that in order to configure the tracer factory, but the documentation earlier suggests replacing the implementation.

File: sdk/lib/opentelemetry/sdk/configurator.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        @adapter_config_map[adapter_name] = config if config
+      end
+
+      # Install all registered instrumentation. Configuration for specific
+      # adapters can be provided with the optional +adapter_config_map+
+      # parameter. Only +use+ or +use_all+, but not both when installing
+      # instrumentation. A call to +use+ after +use_all+ will result in an
+      # exception.
+      #
+      # @param [Hash<String,Hash>] adapter_config_map A map with string keys
```
Comment (fbogsany) [inline sdk/lib/opentelemetry/sdk/configurator.rb]: optional

File: sdk/lib/opentelemetry/sdk.rb (hunk 2 above)
Comment (mwear) [reply sdk/lib/opentelemetry/sdk.rb]: The constant resolution will happen at the point of call. I can move the require if that'd make folks more comfortable.

File: sdk/lib/opentelemetry/sdk/configurator.rb (hunk 3 above)
Comment (mwear) [reply sdk/lib/opentelemetry/sdk/configurator.rb]: I suppose this doesn't make a whole lot of sense. It seems like a user would be using a different SDK, or a hybrid SDK of sorts. In either case, we don't have to support that for the default SDK, and I don't think we should, unless a use case becomes apparent.

File: sdk/lib/opentelemetry/sdk.rb (hunk 2 above)
Comment (fbogsany) [reply sdk/lib/opentelemetry/sdk.rb]: No, this is fine - just wanted to know 😄

File: sdk/lib/opentelemetry/sdk/configurator.rb [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      end
+
+      def configure_span_processors
+        processors = @span_processors.empty? ? [default_span_processor] : @span_processors
+        processors.each { |p| @tracer_factory.add_span_processor(p) }
+      end
+
+      def default_span_processor
+        Trace::Export::SimpleSpanProcessor.new(
+          Trace::Export::ConsoleSpanExporter.new
```
Comment (elskwid) [inline sdk/lib/opentelemetry/sdk/configurator.rb]: This is nice!

**Comment by mwear** [comment]
Relevant to: https://github.com/open-telemetry/opentelemetry-specification/issues/390

**Pull request merged**